embassy-stm32/src/dcmi.rs

```diff
@@ -0,0 +1,393 @@
+        d9: impl Unborrow<Target = impl D9Pin> + 'd,
+        d10: impl Unborrow<Target = impl D10Pin> + 'd,
+        d11: impl Unborrow<Target = impl D11Pin> + 'd,
+        d12: impl Unborrow<Target = impl D12Pin> + 'd,
+        d13: impl Unborrow<Target = impl D13Pin> + 'd,
+        v_sync: impl Unborrow<Target = impl VSyncPin> + 'd,
+        h_sync: impl Unborrow<Target = impl HSyncPin> + 'd,
+        pixclk: impl Unborrow<Target = impl PixClkPin> + 'd,
+    ) -> Self {
+        T::reset();
+        T::enable();
+
+        unborrow!(
+            dma, irq, d0, d1, d2, d3, d4, d5, d6, d7, d8, d9, d10, d11, d12, d13, v_sync, h_sync,
+            pixclk
+        );
+
+        d0.configure();
+        d1.configure();
+        d2.configure();
+        d3.configure();
+        d4.configure();
+        d5.configure();
+        d6.configure();
+        d7.configure();
+        d8.configure();
+        d9.configure();
+        d10.configure();
+        d11.configure();
+        d12.configure();
+        d13.configure();
+
+        v_sync.configure();
+        h_sync.configure();
+        pixclk.configure();
+
+        let edm = match (
+            d8.pin().is_some(),
+            d9.pin().is_some(),
+            d10.pin().is_some(),
+            d11.pin().is_some(),
+            d12.pin().is_some(),
+            d13.pin().is_some(),
+        ) {
+            (true, true, true, true, true, true) => 0b11, // 14 bits
+            (true, true, true, true, false, false) => 0b10, // 12 bits
+            (true, true, false, false, false, false) => 0b01, // 10 bits
+            (false, false, false, false, false, false) => 0b00, // 8 bits
+            _ => {
+                #[cfg(feature = "defmt")]
+                defmt::panic!("Invalid pin configuration.");
+                #[cfg(not(feature = "defmt"))]
+                core::panic!("Invalid pin configuration.");
+            }
+        };
+
+        irq.set_handler(Self::on_interrupt);
+        irq.unpend();
+        irq.enable();
+
+        unsafe {
+            T::regs().cr().modify(|r| {
+                r.set_cm(true); // disable continuous mode (snapshot mode)
+                r.set_ess(use_embedded_synchronization);
+                r.set_pckpol(pixclk_polarity == PixelClockPolarity::RisingEdge);
+                r.set_vspol(vsync_level == VSyncDataInvalidLevel::High);
+                r.set_hspol(hsync_level == HSyncDataInvalidLevel::High);
+                r.set_fcrc(0x00); // capture every frame
+                r.set_edm(edm); // extended data mode
+            });
+            T::regs().ier().modify(|r| {
+                r.set_ovr_ie(true);
+                r.set_frame_ie(true);
+            });
+        }
+
+        Self {
+            phantom: PhantomData,
+            dma,
+        }
+    }
+
+    unsafe fn on_interrupt(_: *mut ()) {
+        let ris = T::regs().ris().read();
+        if ris.err_ris() {
+            #[cfg(feature = "defmt")]
+            defmt::error!("DCMI: Error.");
```

Comment:
just do `error!("DCMI: Error.");`. The macros in `src/fmt.rs` will automatically compile to using `defmt`, `log` or nothing depending on what's enabled.